File: backend/service/seed-demo.mjs
// Seed a running service with enough real people to walk the six workspaces.
//
// WHAT THIS IS NOT: canned content. Every row it creates is created through the
// real path — the bootstrap ceremony, then the administrator's own endpoints —
// so nothing here can produce a state the system could not reach on its own.
// That is the difference between a seed and a mockup, and it is the whole
// reason the v4 concept's invented rows are not imported anywhere.
//
// It creates six accounts, one per role, so that somebody can sign in as each
// and see their workspace. It creates no deals, no tickets and no clauses:
// those are acts for the people in the system to perform, and an empty
// workspace showing its honest empty state is the correct first impression.
//
//   node service/seed-demo.mjs --data ./cw-data
//   node service/seed-demo.mjs                  (in-memory; prints and exits)

import { Db } from './db.mjs';

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    if (!argv[i].startsWith('--')) continue;
    const k = argv[i].slice(2);
    out[k] = argv[i + 1] && !argv[i + 1].startsWith('--') ? argv[++i] : 'true';
  }
  return out;
}
const args = parseArgs(process.argv.slice(2));

const OWNER = 'owner@clausewerk';
const PEOPLE = [
  ['a.okafor@clausewerk', 'Ada Okafor',   'Operations',  'administrator'],
  ['r.vance@clausewerk',  'Rae Vance',    'Legal',       'legal_admin'],
  ['p.nkemi@clausewerk',  'Pat Nkemi',    'Legal',       'legal_reviewer'],
  ['d.buyer@clausewerk',  'Dana Buyer',   'Procurement', 'requester'],
  ['t.imani@clausewerk',  'Tunde Imani',  'Assurance',   'auditor'],
  ['s.reed@clausewerk',   'Sam Reed',     'Supplier',    'viewer'],
];
const [ADMIN, LEGAL] = [PEOPLE[0][0], PEOPLE[1][0]];

const db = await Db.open({ dataDir: args.data });

// The ceremony creates the first two. It refuses if it has run before, so this
// script is safe to re-run against an existing data directory — it will say so
// and stop rather than pretend.
try {
  await db.asPerson(OWNER, 'viewer', () => {});   // no-op; keeps the shape honest
} catch { /* ignore */ }

const already = await db.lookUpIdentity(({ query }) =>
  query(`select count(*)::int n from cw.account`));

if (already[0].n > 0) {
  console.log(`this database already has ${already[0].n} account(s); nothing to do.`);
  console.log('the bootstrap ceremony happens once, and accounts are revoked rather than deleted.');
  await db.close();
  process.exit(0);
}

// The ceremony needs owner rights, which the service deliberately cannot reach.
// This script is installation, not a request, so it goes through Db's migrate
// path — the same asymmetry the service tests rely on.
await db.bootstrap(OWNER, PEOPLE[0], PEOPLE[1]);

// The remaining four, created and granted by the Administrator, through exactly
// the acts the console will perform in WP-U08.
await db.asPerson(ADMIN, 'administrator', async ({ query }) => {
  for (const [person, name, unit, role] of PEOPLE.slice(2)) {
    await query(`insert into cw.account (person,display_name,unit,role,created_by)
                 values ($1,$2,$3,$4,current_setting('cw.actor'))`, [person, name, unit, role]);
    await query(`insert into cw.role_grant (action,person,role,reason)
                 values ('granted',$1,$2,'seeded for the walkthrough')`, [person, role]);
  }
});

// Pat holds a LEGAL role, so their grant confers nothing until Rae accepts it.
// Doing this here rather than silently is the point: a walkthrough that skipped
// the countersign would leave the reviewer unable to sign in, and somebody would
// "fix" it by removing the rule.
await db.asPerson(LEGAL, 'legal_admin', async ({ query }) => {
  const g = await query(`select grant_id from cw.role_grant
                         where person = 'p.nkemi@clausewerk' and action = 'granted'`);
  await query(`insert into cw.role_grant (action,person,role,grant_ref)
               values ('countersigned','p.nkemi@clausewerk','legal_reviewer',$1)`,
    [g[0].grant_id]);
});

const effective = await db.lookUpIdentity(({ query }) =>
  query(`select person, display_name, role from cw.effective_role order by role`));

console.log('six people, one per role. Sign in as any of them:\n');
for (const p of effective)
  console.log(`  ${p.role.padEnd(15)} ${p.person.padEnd(24)} ${p.display_name}`);

console.log(`
No deals, tickets or clauses were created. Those are acts for the people above
to perform, and an empty workspace showing its honest empty state is the correct
first impression — a seeded system that looks busy is a demo, not a system.`);

await db.close();

File: backend/doorway/seed_demo.py
"""Seed a database with enough real people to walk the six workspaces.

    python -m doorway.seed_demo

WHAT THIS IS NOT: canned content. Every row it creates is created through the
real path — the bootstrap ceremony, then the Administrator's own acts, through
the doorway, as themselves. Nothing here can produce a state the system could not
reach on its own. That is the difference between a seed and a mockup, and it is
the whole reason the v4 concept's invented rows are imported nowhere.

It creates six accounts, one per role, so somebody can sign in as each and see
their workspace. It creates no deals, no tickets and no clauses: those are acts
for the people in the system to perform, and an empty workspace showing its
honest empty state is the correct first impression. A seeded system that looks
busy is a demo, not a system.

Ported from `backend/service/seed-demo.mjs` on 2026-07-26.
"""

from __future__ import annotations

import os
import sys

import psycopg

from doorway.db import Database
from doorway.setup import OWNER_URL, prepare

APP_URL = os.environ.get(
    "CW_DATABASE_URL", "postgresql://cw_app:clausewerk-dev@localhost:5432/clausewerk")

OWNER = "owner@clausewerk"

PEOPLE = [
    ("a.okafor@clausewerk", "Ada Okafor", "Operations", "administrator"),
    ("r.vance@clausewerk", "Rae Vance", "Legal", "legal_admin"),
    ("p.nkemi@clausewerk", "Pat Nkemi", "Legal", "legal_reviewer"),
    ("d.buyer@clausewerk", "Dana Buyer", "Procurement", "requester"),
    ("t.imani@clausewerk", "Tunde Imani", "Assurance", "auditor"),
    ("s.reed@clausewerk", "Sam Reed", "Supplier", "viewer"),
]

ADMIN = PEOPLE[0][0]
LEGAL = PEOPLE[1][0]


def seed(owner_url: str = OWNER_URL, app_url: str = APP_URL) -> list[dict]:
    """Run the ceremony and create the other four. Returns who can sign in."""
    prepare(owner_url=owner_url)

    with psycopg.connect(owner_url, autocommit=True) as owner:
        already = owner.execute("select count(*) from cw.account").fetchone()[0]
        if already:
            print(f"this database already has {already} account(s); nothing to do.")
            print("the bootstrap ceremony happens once, and accounts are revoked "
                  "rather than deleted.")
            return []

        # The ceremony needs owner rights, which the doorway deliberately cannot
        # reach. This script is installation, not a request — the same asymmetry
        # migrate.py rests on.
        # The ceremony takes ONE unit and gives it to both accounts it creates —
        # see cw.bootstrap's p_unit. It cannot put the first Administrator in
        # Operations and the first Legal admin in Legal, and that is a real limit
        # rather than an oversight: at the moment of the ceremony there is no
        # Administrator yet to say where anybody sits.
        #
        # So the Administrator's unit goes in here, and Rae's is corrected below
        # by the Administrator, as an ordinary recorded act. The alternative —
        # writing it straight in as the owner — would put a row in the system
        # that the system could not have reached on its own, which is the one
        # thing this file exists not to do.
        owner.execute(
            "select cw.bootstrap(%s,%s,%s,%s,%s,%s)",
            (OWNER, PEOPLE[0][0], PEOPLE[0][1], PEOPLE[1][0], PEOPLE[1][1],
             PEOPLE[0][2]),
        )

    db = Database(app_url)
    try:
        # The remaining four, created and granted by the Administrator, through
        # exactly the acts the console performs.
        with db.as_person(ADMIN, "administrator") as request:
            # The correction the ceremony could not make. An ordinary update by
            # the Administrator, audited like any other.
            request.write_one(
                "update cw.account set unit = %s where person = %s",
                (PEOPLE[1][2], PEOPLE[1][0]))

            for person, name, unit, role in PEOPLE[2:]:
                request.write_one(
                    "insert into cw.account (person, display_name, unit, role, "
                    "created_by) values (%s,%s,%s,%s,current_setting('cw.actor'))",
                    (person, name, unit, role))
                request.write_one(
                    "insert into cw.role_grant (action, person, role, reason) "
                    "values ('granted',%s,%s,'seeded for the walkthrough')",
                    (person, role))

        # Pat holds a LEGAL role, so their grant confers nothing until Rae
        # accepts it. Doing this here rather than silently is the point: a
        # walkthrough that skipped the countersign would leave the reviewer
        # unable to sign in, and somebody would "fix" it by removing the rule.
        with db.as_person(LEGAL, "legal_admin") as request:
            grant = request.one(
                "select grant_id from cw.role_grant "
                "where person = %s and action = 'granted'", ("p.nkemi@clausewerk",))
            request.write_one(
                "insert into cw.role_grant (action, person, role, grant_ref) "
                "values ('countersigned','p.nkemi@clausewerk','legal_reviewer',%s)",
                (grant[0],))

        with db.as_person(ADMIN, "administrator") as request:
            return request.rows(
                "select person, display_name, role from cw.effective_role "
                "order by role")
    finally:
        db.close()


def main() -> int:
    effective = seed()
    if not effective:
        return 0

    print("six people, one per role. Sign in as any of them:\n")
    for row in effective:
        print(f"  {row['role']:<15} {row['person']:<24} {row['display_name']}")

    print("""
No deals, tickets or clauses were created. Those are acts for the people above
to perform, and an empty workspace showing its honest empty state is the correct
first impression — a seeded system that looks busy is a demo, not a system.""")
    return 0


if __name__ == "__main__":
    sys.exit(main())
